waku/node/peer_manager/peer_manager.nim

```diff
@@ -906,6 +910,55 @@ proc prunePeerStoreLoop(pm: PeerManager) {.async.} =
     pm.prunePeerStore()
     await sleepAsync(PrunePeerStoreInterval)
 
+proc calculateTopicHealth(wakuRelay: WakuRelay, topic: string): TopicHealth =
+  let numPeersInMesh = wakuRelay.getNumPeersInMesh(topic).valueOr:
+    error "Could not calculate topic health", topic = topic, error = error
+    return TopicHealth.UNHEALTHY
+
+  if numPeersInMesh < 1:
+    return TopicHealth.UNHEALTHY
+  elif numPeersInMesh < wakuRelay.parameters.dLow:
+    return TopicHealth.MINIMALLY_HEALTHY
+  return TopicHealth.SUFFICIENTLY_HEALTHY
+
+proc updateTopicsHealth(pm: PeerManager): Future[void] =
+  if pm.wakuRelay.isNil():
+    return
+
+  var futs = newSeq[Future[void]]()
+  for topic in toSeq(pm.wakuRelay.topics.keys):
+    let
+      oldHealth = pm.topicsHealth.getOrDefault(topic)
+      currentHealth = pm.wakuRelay.calculateTopicHealth(topic)
+
+    if oldHealth == currentHealth:
+      continue
+
+    pm.topicsHealth[topic] = currentHealth
+    if not pm.onTopicHealthChange.isNil():
+      let fut = pm.onTopicHealthChange(topic, currentHealth)
+      if not fut.completed(): # Fast path for successful sync handlers
+        futs.add(fut)
+
+  if futs.len() > 0:
+    proc waiter(): Future[void] {.async.} =
+      # slow path - we have to wait for the handlers to complete
+      try:
+        futs = await allFinished(futs)
+      except CancelledError:
+        # propagate cancellation
+        for fut in futs:
+          if not (fut.finished):
+            await fut.cancelAndWait()
+
+      # check for errors in futures
+      for fut in futs:
+        if fut.failed:
+          let err = fut.readError()
+          warn "Error in health change handler", description = err.msg
+
+    return waiter()
```

Comment:
This part looks horrible honestly - copied it from nim-libp2p's handling of Relay callback events

https://github.com/vacp2p/nim-libp2p/blob/fd26f93b80cfeecd4e6f9c84eb7a15a3be26c1ee/libp2p/protocols/pubsub/pubsub.nim#L379-L408

It's in part necessary because the callbacks configured can be async